### DISCUSSION TITLE
Adding Role Detection and Structural Holes into NetworkX

### DISCUSSION THREAD

**Discussion by BradKML** [networkx/networkx#4673] [Ideas]
Roles in networks are important for finding out how communities operate.
[Some use this for finding out who the leaders and references are within a community.](https://www.groundai.com/project/interest-communities-and-flow-roles-in-directed-networks-the-twitter-network-of-the-uk-riots)
Take a look at https://github.com/GiulioRossetti/cdlib/issues/137 and https://github.com/GiulioRossetti/cdlib/issues/149

**Comment by dschult** [comment]
NetworkX has at least some functionality for `structural_holes` in `networkx/algorithms/structuralholes.py`.
I'm sure that more could be done here.

**Reply by BradKML** [reply]
I am more or less eyeing Role Detection or Role Discovery since those are more useful in social media analysis.

**Comment by BradKML** [comment]
@dschult what do you think about this? Should this be added to NetworkX? It is similar to Link Analysis, but for Roles.
https://github.com/i-eat-bread/RoleMatch-Measure
https://etd.ohiolink.edu/apexprod/rws_etd/send_file/send?accession=kent1345064156&disposition=inline

**Reply by dschult** [reply]
This code looks like a first step toward creating a matrix. But I don't see that it actually does anything at this oint. But it is a start.  Spend some time figuring out how this stuff works. Construct some examples and see whether they measure something interesting in a case where you know the answer.  Once you know enough to build a functioning prototype and rewrite it once or twice you should be in good shape :}
